src/components/application_manager/src/resumption/resumption_data_processor_impl.cc

```diff
@@ -79,40 +79,54 @@ void ResumptionDataProcessorImpl::Restore(
     ResumeCtrl::ResumptionCallBack callback) {
   SDL_LOG_AUTO_TRACE();
 
-  if (!HasDataToRestore(saved_app)) {
+  bool restore_hmi_data = HasHMIDataToRestore(saved_app);
+  bool restore_local_data = HasLocalDataToRestore(saved_app);
+  if (!restore_hmi_data && !restore_local_data) {
     SDL_LOG_DEBUG("No data to restore, resumption is successful");
     callback(mobile_apis::Result::SUCCESS, "Data resumption succesful");
     return;
   }
 
+  const auto app_id = application->app_id();
+  if (!restore_hmi_data && restore_local_data) {
+    AddSubscriptions(application, saved_app);
+    SDL_LOG_DEBUG("No requests to HMI for " << app_id
+                                            << ", resumption is successful");
+    callback(mobile_apis::Result::SUCCESS, "Data resumption successful");
+    return;
+  }
+
+  register_callbacks_lock_.AcquireForWriting();
+  register_callbacks_[app_id] = callback;
+  register_callbacks_lock_.Release();
+
   AddFiles(application, saved_app);
   AddSubmenus(application, saved_app);
   AddCommands(application, saved_app);
   AddChoicesets(application, saved_app);
   SetGlobalProperties(application, saved_app);
   AddSubscriptions(application, saved_app);
   AddWindows(application, saved_app);
+}
 
-  resumption_status_lock_.AcquireForReading();
-  const auto app_id = application->app_id();
-  bool is_requests_list_empty = true;
-  if (resumption_status_.find(app_id) != resumption_status_.end()) {
-    is_requests_list_empty =
-        resumption_status_[app_id].list_of_sent_requests.empty();
-  }
-  resumption_status_lock_.Release();
+bool ResumptionDataProcessorImpl::HasLocalDataToRestore(
+    const smart_objects::SmartObject& saved_app) const {
+  SDL_LOG_AUTO_TRACE();
 
-  if (!is_requests_list_empty) {
-    sync_primitives::AutoWriteLock lock(register_callbacks_lock_);
-    register_callbacks_[app_id] = callback;
-  } else {
-    SDL_LOG_DEBUG("No requests to HMI for " << app_id
-                                            << " , resumption is successful");
-    callback(mobile_apis::Result::SUCCESS, "Data resumption successful");
```

Comment:
@ychernysheva I am still thinking that you should leave this if-else block here.
Let's say, something has happened in `AddCommands` and it takes about 1 second to execute. At this moment, HMI responds to all requests triggered by `AddSubmenus` and processor thinks that resumption has been completed and tries to invoke the callback. Because in your case it was moved above all that staff, processor will be able to find that callback and respond to RAI with RESUME_FAILED and revert submenues. After that, another thread continues to execute the rest of code in `Restore` and sends the rest of resumption requests even after RESUME_FAILED response!
I think that is a reason why callback is added into the map AFTER all resumption requests are sent.
From another side, I understand what is the issue and I propose just to do to like that:
```
if (!is_requests_list_empty) {
  sync_primitives::AutoWriteLock lock(register_callbacks_lock_);
  register_callbacks_[app_id] = callback;
  return;
}

if (IsResumptionSuccessful(app_id)) {
    SDL_LOG_DEBUG("Resumption for app " << app_id << " successful");
    callback(mobile_apis::Result::SUCCESS, "Data resumption successful");
    application_manager_.state_controller().ResumePostponedWindows(app_id);
  } else {
    SDL_LOG_ERROR("Resumption for app " << app_id << " failed");
    callback(mobile_apis::Result::RESUME_FAILED, "Data resumption failed");
    RevertRestoredData(application_manager_.application(app_id));
    application_manager_.state_controller().DropPostponedWindows(app_id);
  }

  EraseAppResumptionData(app_id, function_id, corr_id);
```

As the last several lines of code is the copy of the same actions in `ProcessResponseFromHMI` I suggest to extract this to a separate function and use in both places. In that case SDL will behave as expected because we don't have any guarantees that there will be a strict timings during the resumption process.

Reply:
@AKalinich-Luxoft I agree that we need to leave code block, responsible for `callback` adding, after all resumption requests are sent. But we really need to get rid of next shameful `else` code block, and existence of this code block is a root of all evil in this case.

The one and only purpose of this `else` with condition `is_requests_list_empty`  is sending successful RAI response for non-HMI subscriptions like AppServices and SysCaps in the case when we don't have any HMI subscriptions.

So we really don't need to process it here and can move this code block above, where it cannot cause strange behavior or broke anything.

Reply:
@AKalinich-Luxoft Code block, responsible for callback adding, is left after all resumption requests are sent in f923c29089b198c9dff73c07a1f705164eade5b3